src/Oscoin/Data/Core.hs

```diff
@@ -0,0 +1,502 @@
+                 , (.)
+                 , (<>)
+                 )
+
+import           Oscoin.Crypto.Address (Address)
+import           Oscoin.Crypto.Address.Serialisation (serializeAddress)
+import           Oscoin.Crypto.Blockchain (Height)
+import           Oscoin.Crypto.Blockchain.Block (BlockHash)
+import qualified Oscoin.Crypto.Hash as Crypto
+import           Oscoin.Crypto.PubKey
+                 (HasDigitalSignature, PublicKey, Signature, Signed)
+
+import qualified Codec.Serialise as CBOR
+import           Crypto.Data.Auth.Tree (Tree)
+import qualified Crypto.Data.Auth.Tree as WorldState
+import           Data.ByteArray (ByteArrayAccess)
+import           Data.ByteString (ByteString)
+import qualified Data.Map as Map
+import qualified Data.Set as Set
+import           Numeric.Natural
+
+-------------------------------------------------------------------------------
+-- World State
+-------------------------------------------------------------------------------
+
+-- | The state of the world, materialized from transactions.
+type WorldState c = Tree StateKey (StateVal c)
+
+-- | State lookup key.
+type StateKey = ByteString
+
+-- | State value.
+data StateVal c =
+      AccountVal (Account c)
+    | ProjectVal (Project c)
+    | MemberVal  (Member c)
+    | NatVal     Natural
+
+-- | Like 'Map.adjust', but for 'WorldState'.
+adjust :: (StateVal c -> StateVal c) -> StateKey -> WorldState c -> WorldState c
+adjust f k ws =
+    case WorldState.lookup k ws of
+        Just v  -> WorldState.insert k (f v) ws
+        Nothing -> ws
+
+-- | Like 'Map.alter', but for 'WorldState'.
+alter
+    :: (Maybe (StateVal c) -> Maybe (StateVal c))
+    -> StateKey
+    -> WorldState c
+    -> WorldState c
+alter f k ws =
+    case f (WorldState.lookup k ws) of
+        Just v  -> WorldState.insert k v ws
+        Nothing -> WorldState.delete k ws
+
+-------------------------------------------------------------------------------
+-- Core types
+-------------------------------------------------------------------------------
+
+-- | A balance of oscoin in the smallest denomination.
+type Balance = Word64
+
+-- | An account nonce to prevent replay attacks.
+type Nonce = Word64
+
+-- | An account which holds a balance.
+data Account c = Account
+    { accountAddr    :: Address c
+    -- ^ The account identifier.
+    , accountBalance :: Balance
+    -- ^ The oscoin balance.
+    , accountNonce   :: Nonce
+    -- ^ The nonce is equal to the number of transactions
+    -- made from this account.
+    }
+
+mkAccount :: Address c -> Account c
+mkAccount addr = Account
+    { accountAddr    = addr
+    , accountBalance = 0
+    , accountNonce   = 0
+    }
+
+-- | Increase the account balance.
+deposit :: Balance -> Account c -> Account c
+deposit bal acc = acc { accountBalance = accountBalance acc + bal }
+
+-- | Decrease the account balance.
+withdraw :: Balance -> Account c -> Account c
+withdraw bal acc = acc { accountBalance = accountBalance acc - bal }
```

Comment:
We need to check against underflow here.

Reply:
2f5fc95